swift/ql/src/queries/Security/CWE-135/StringLengthConflation.ql

```diff
@@ -11,5 +11,78 @@
  */
 
 import swift
+import codeql.swift.dataflow.DataFlow
+import DataFlow::PathGraph
 
-select "TODO"
+class StringLengthConflationConfiguration extends DataFlow::Configuration {
+  StringLengthConflationConfiguration() { this = "StringLengthConflationConfiguration" }
+
+  override predicate isSource(DataFlow::Node node, string flowstate) {
+    // result of a call to to `String.count`
+    exists(MemberRefExpr member |
+      member.getBaseExpr().getType().toString() = "String" and // TODO: use of toString
+      member.getMember().toString() = "count" and // TODO: use of toString
+      node.asExpr() = member and
+      flowstate = "String"
+    )
+  }
+
+  override predicate isSink(DataFlow::Node node, string flowstate) {
+    // arguments to method calls...
+    exists(
+      string className, string methodName, string argName, ClassDecl c, AbstractFunctionDecl f,
+      CallExpr call, int arg
+    |
+      (
+        // `NSRange.init`
+        className = "NSRange" and
+        methodName = "init" and
+        argName = ["location", "length"]
+        or
+        // `NSString.character`
+        className = ["NSString", "NSMutableString"] and
+        methodName = "character" and
+        argName = "at"
+        or
+        // `NSString.character`
+        className = ["NSString", "NSMutableString"] and
+        methodName = "substring" and
+        argName = ["from", "to"]
+        or
+        // `NSMutableString.insert`
+        className = "NSMutableString" and
+        methodName = "insert" and
+        argName = "at"
+      ) and
+      c.toString() = className and // TODO: use of toString
+      c.getAMember() = f and // TODO: will this even work if its defined in a parent class?
+      call.getFunction().(ApplyExpr).getFunction().(DeclRefExpr).getDecl() = f and
```

Comment:
If you make sure that this branch includes https://github.com/github/codeql/pull/9623/commits/e2478d1106bb464f3171b33a0c1375e17eabd48d then I think this should be equivalent
```suggestion
      call.getStaticTarget() = f and
```

Reply:
Also doesn't work right now, I'll check I have that library fix that made this work on constructors.

Reply:
Please do! It certainly appeared to do that when looking at the [diff](https://github.com/github/codeql/commit/e2478d1106bb464f3171b33a0c1375e17eabd48d#diff-2e7af3b9876580e16b61a25d4f8e824927a2e2be4ac855964b0d27cad4101df7R47) 🤔.

Reply:
Oh, wait. Yeah, I see why my change didn't work: you're writing is equivalent to:
```swift
call.getFunction().(ApplyExpr).getStaticTarget() = f
```
Now, why you need that second layer of indirection isn't clear to me.

Reply:
Yep, that works (change made).  I assume the second level of indirection is because `s.count` itself is a call to an implicit `ApplyExpr` on `s` that returns the member function `count`???  I'm not sure at all though.

Reply:
I think maybe a chunk of this logic should be made to look something like this:
```
...
c.getMember(methodName) = f and
call.get???Target() = f and
...
```

Reply:
Ahhh... Yeah, now I see what's going on. Gosh, this is a bit annoying:

For this piece of code:
```swift
let ns = NSString(string: s)
```
There are two function calls. This is because it's represented as `(NSString.init(this))(string: s)`. So the inner-most function supplies the `this` argument, and returns the function to be invokes, and the outer-most function actually does the call.

We should figure out what the best way of navigating this in QL is. For now, it looks like you've worked around the issue quite well here 👍.